### PR TITLE
Introduce a mechanism to build all Kubeflow images

### PR DESCRIPTION
**Pull request by apo-ger**
Introduce a mechanism to build all Kubeflow images. We use a top-level Makefile to: 
- call the existing sub-Makefiles of the different components for building/pushing the images.
- use a specific REGISTRY to retag & push the images.

For building example notebook-server images, we:
- Create a sub-Makefile for each of the notebook servers, in each folder:
  - Each sub-makefile will be responsible for building the required bases (we don't expect them to be already present) and use ARGs for passing them on the appropriate Dockerfiles.
- Create a central-makefile under example-notebook-servers/ dir to call each sub-Makefile from above.

We modified every sub-Makefile to ensure that:
- we don't have a registry prefix when building images
- use TAG ?= $(shell git describe --tags --dirty)

### CODE REVIEW

File: components/access-management/Makefile [hunk 1]
```diff
@@ -1,30 +1,31 @@
...
+
+docker-build:
 	docker build -t $(IMG):$(TAG) .
 
+docker-push: 
+	docker push $(IMG):$(TAG)
+
+image: docker-build docker-push 
+
 build-gcb:
```
Comment (kimwnasptd) [inline components/access-management/Makefile:26]: Can you also remove this rule? We don't use `build-gcb` anywhere in the project.

It's a good chance to clean up the project

File: components/admission-webhook/Makefile [hunk 2]
```diff
@@ -1,12 +1,19 @@
...
 
 docker-build:
+	docker build -t ${IMG}:${TAG} -f Dockerfile .
+
+docker-push: 
+	docker push ${IMG}:${TAG}
+
+image: docker-build docker-push
+
+build-gcr:
```
Comment (kimwnasptd) [inline components/admission-webhook/Makefile:16]: Same here, let's completely remove GCR specific commands

File: components/notebook-controller/Makefile [hunk 3]
```diff
@@ -77,12 +80,17 @@ run: manifests generate fmt vet ## Run a controller from your host.
 
 .PHONY: docker-build
 docker-build: test ## Build docker image with the manager.
+=======
```
Comment (kimwnasptd) [inline components/notebook-controller/Makefile:83]: I think there's some conflict resolution that needs to happen in this file

File: components/tensorboard-controller/Makefile [hunk 4]
```diff
@@ -40,9 +39,16 @@ help: ## Display this help.
 
 ##@ Development
 
+<<<<<<< HEAD
```
Comment (kimwnasptd) [inline components/tensorboard-controller/Makefile:42]: Some conflict resolution needs to happen in this file as well

**Review by kimwnasptd — COMMENTED** [review]
Adding some review comments from a first pass. Will add another batch regarding the notebook-server images specifically

File: components/example-notebook-servers/codeserver/Dockerfile [hunk 5]
```diff
@@ -1,10 +1,11 @@
-FROM public.ecr.aws/j1r0q0g6/notebooks/notebook-servers/base:master-6c41e72b
+ARG BASE_IMG=public.ecr.aws/j1r0q0g6/notebooks/notebook-servers/base:master-6c41e72b
```
Comment (kimwnasptd) [inline components/example-notebook-servers/codeserver/Dockerfile:1]: Same for here as well:
1. `BASE_IMG=<base>`
2. A comment to instruct users to use the makefile to build the image

File: components/example-notebook-servers/jupyter-pytorch/cpu.Dockerfile [hunk 6]
```diff
@@ -1,4 +1,5 @@
-FROM public.ecr.aws/j1r0q0g6/notebooks/notebook-servers/jupyter:master-434b10ab
+ARG BASE_IMG=public.ecr.aws/j1r0q0g6/notebooks/notebook-servers/jupyter:master-434b10ab
```
Comment (kimwnasptd) [inline components/example-notebook-servers/jupyter-pytorch/cpu.Dockerfile:1]: same

File: components/example-notebook-servers/jupyter-pytorch/cuda.Dockerfile [hunk 7]
```diff
@@ -1,4 +1,5 @@
-FROM public.ecr.aws/j1r0q0g6/notebooks/notebook-servers/jupyter:master-434b10ab
+ARG BASE_IMG=public.ecr.aws/j1r0q0g6/notebooks/notebook-servers/jupyter:master-434b10ab
```
Comment (kimwnasptd) [inline components/example-notebook-servers/jupyter-pytorch/cuda.Dockerfile:1]: same

File: components/example-notebook-servers/jupyter-scipy/Dockerfile [hunk 8]
```diff
@@ -1,4 +1,5 @@
-FROM public.ecr.aws/j1r0q0g6/notebooks/notebook-servers/jupyter:master-434b10ab
+ARG BASE_IMG=public.ecr.aws/j1r0q0g6/notebooks/notebook-servers/jupyter:master-434b10ab
```
Comment (kimwnasptd) [inline components/example-notebook-servers/jupyter-scipy/Dockerfile:1]: same

File: components/example-notebook-servers/jupyter-tensorflow-full/cpu.Dockerfile [hunk 9]
```diff
@@ -1,4 +1,5 @@
-FROM public.ecr.aws/j1r0q0g6/notebooks/notebook-servers/jupyter-tensorflow:master-e9324d39
+ARG BASE_IMG=public.ecr.aws/j1r0q0g6/notebooks/notebook-servers/jupyter-tensorflow:master-e9324d39
```
Comment (kimwnasptd) [inline components/example-notebook-servers/jupyter-tensorflow-full/cpu.Dockerfile:1]: same

File: components/example-notebook-servers/jupyter-tensorflow-full/cuda.Dockerfile [hunk 10]
```diff
@@ -1,5 +1,5 @@
-FROM public.ecr.aws/j1r0q0g6/notebooks/notebook-servers/jupyter-tensorflow-cuda:master-e9324d39
-
+ARG BASE_IMG=public.ecr.aws/j1r0q0g6/notebooks/notebook-servers/jupyter-tensorflow-cuda:master-e9324d39
```
Comment (kimwnasptd) [inline components/example-notebook-servers/jupyter-tensorflow-full/cuda.Dockerfile:1]: same

File: components/example-notebook-servers/jupyter-tensorflow/cpu.Dockerfile [hunk 11]
```diff
@@ -1,4 +1,5 @@
-FROM public.ecr.aws/j1r0q0g6/notebooks/notebook-servers/jupyter:master-434b10ab
+ARG BASE_IMG=public.ecr.aws/j1r0q0g6/notebooks/notebook-servers/jupyter:master-434b10ab
```
Comment (kimwnasptd) [inline components/example-notebook-servers/jupyter-tensorflow/cpu.Dockerfile:1]: same

File: components/example-notebook-servers/jupyter-tensorflow/cuda.Dockerfile [hunk 12]
```diff
@@ -1,4 +1,5 @@
-FROM public.ecr.aws/j1r0q0g6/notebooks/notebook-servers/jupyter:master-434b10ab
+ARG BASE_IMG=public.ecr.aws/j1r0q0g6/notebooks/notebook-servers/jupyter:master-434b10ab
```
Comment (kimwnasptd) [inline components/example-notebook-servers/jupyter-tensorflow/cuda.Dockerfile:1]: same

File: components/example-notebook-servers/jupyter/Dockerfile [hunk 13]
```diff
@@ -1,4 +1,5 @@
-FROM public.ecr.aws/j1r0q0g6/notebooks/notebook-servers/base:master-6c41e72b
+ARG BASE_IMG=public.ecr.aws/j1r0q0g6/notebooks/notebook-servers/base:master-6c41e72b
```
Comment (kimwnasptd) [inline components/example-notebook-servers/jupyter/Dockerfile:1]: same

File: components/example-notebook-servers/rstudio-tidyverse/Dockerfile [hunk 14]
```diff
@@ -1,4 +1,5 @@
-FROM public.ecr.aws/j1r0q0g6/notebooks/notebook-servers/rstudio:master-434b10ab
+ARG BASE_IMG=public.ecr.aws/j1r0q0g6/notebooks/notebook-servers/rstudio:master-434b10ab
```
Comment (kimwnasptd) [inline components/example-notebook-servers/rstudio-tidyverse/Dockerfile:1]: same

File: components/example-notebook-servers/rstudio/Dockerfile [hunk 15]
```diff
@@ -1,4 +1,5 @@
-FROM public.ecr.aws/j1r0q0g6/notebooks/notebook-servers/base:master-6c41e72b
+ARG BASE_IMG=public.ecr.aws/j1r0q0g6/notebooks/notebook-servers/base:master-6c41e72b
```
Comment (kimwnasptd) [inline components/example-notebook-servers/rstudio/Dockerfile:1]: same

**Review by kimwnasptd — COMMENTED** [review]
Regarding the notebook server images, let's completely remove the previous logic of using already built images as bases. This overcomplicates everything and expects updates to happen in two steps.

Instead we should have makefiles everywhere that:
1. Can build any base image
2. If an image requires another notebook base, then we first build that one using its makefile, and then use it as docker `ARG` for building the next one

So this way we can build everything with the latest code and don't have to do updates in batches

**Comment by apo-ger** [comment]
@kimwnasptd 

> Adding some review comments from a first pass. Will add another batch regarding the notebook-server images specifically

Fixed!

> Regarding the notebook server images, let's completely remove the previous logic of using already built images as bases. This overcomplicates everything and expects updates to happen in two steps.
> 
> Instead we should have makefiles everywhere that:
> 
> 1. Can build any base image
> 2. If an image requires another notebook base, then we first build that one using its makefile, and then use it as docker `ARG` for building the next one
> 
> So this way we can build everything with the latest code and don't have to do updates in batches
Fixed!

Fixes are included in distinct commits

File: components/access-management/Makefile (hunk 1 above)
Comment (apo-ger) [reply components/access-management/Makefile]: Done!

File: components/admission-webhook/Makefile (hunk 2 above)
Comment (apo-ger) [reply components/admission-webhook/Makefile]: Done!

File: components/notebook-controller/Makefile (hunk 3 above)
Comment (apo-ger) [reply components/notebook-controller/Makefile]: Done!

File: components/tensorboard-controller/Makefile (hunk 4 above)
Comment (apo-ger) [reply components/tensorboard-controller/Makefile]: Done!

File: components/example-notebook-servers/codeserver-python/Dockerfile [hunk 16]
```diff
@@ -1,4 +1,5 @@
-FROM public.ecr.aws/j1r0q0g6/notebooks/notebook-servers/codeserver:master-434b10ab
+ARG BASE_IMG=public.ecr.aws/j1r0q0g6/notebooks/notebook-servers/codeserver:master-434b10ab
```
[the remark this answers is not in the document]
Comment (apo-ger) [reply components/example-notebook-servers/codeserver-python/Dockerfile]: Fixed!

File: components/example-notebook-servers/codeserver/Dockerfile (hunk 5 above)
Comment (apo-ger) [reply components/example-notebook-servers/codeserver/Dockerfile]: Fixed!

File: components/example-notebook-servers/jupyter-pytorch-full/cpu.Dockerfile [hunk 17]
```diff
@@ -1,4 +1,5 @@
-FROM public.ecr.aws/j1r0q0g6/notebooks/notebook-servers/jupyter-pytorch:master-c7ed4a32
+ARG BASE_IMG=public.ecr.aws/j1r0q0g6/notebooks/notebook-servers/jupyter-pytorch:master-c7ed4a32
```
[the remark this answers is not in the document]
Comment (apo-ger) [reply components/example-notebook-servers/jupyter-pytorch-full/cpu.Dockerfile]: Fixed!

File: components/example-notebook-servers/jupyter-pytorch-full/cuda.Dockerfile [hunk 18]
```diff
@@ -1,4 +1,5 @@
-FROM public.ecr.aws/j1r0q0g6/notebooks/notebook-servers/jupyter-pytorch-cuda:master-c7ed4a32
+ARG BASE_IMG=public.ecr.aws/j1r0q0g6/notebooks/notebook-servers/jupyter-pytorch-cuda:master-c7ed4a32
```
[the remark this answers is not in the document]
Comment (apo-ger) [reply components/example-notebook-servers/jupyter-pytorch-full/cuda.Dockerfile]: Fixed!

File: components/example-notebook-servers/jupyter-pytorch/cpu.Dockerfile (hunk 6 above)
Comment (apo-ger) [reply components/example-notebook-servers/jupyter-pytorch/cpu.Dockerfile]: Fixed!

File: components/example-notebook-servers/jupyter-pytorch/cuda.Dockerfile (hunk 7 above)
Comment (apo-ger) [reply components/example-notebook-servers/jupyter-pytorch/cuda.Dockerfile]: Fixed!

File: components/example-notebook-servers/jupyter-scipy/Dockerfile (hunk 8 above)
Comment (apo-ger) [reply components/example-notebook-servers/jupyter-scipy/Dockerfile]: Fixed!

File: components/example-notebook-servers/jupyter-tensorflow-full/cpu.Dockerfile (hunk 9 above)
Comment (apo-ger) [reply components/example-notebook-servers/jupyter-tensorflow-full/cpu.Dockerfile]: Fixed!

File: components/example-notebook-servers/jupyter-tensorflow-full/cuda.Dockerfile (hunk 10 above)
Comment (apo-ger) [reply components/example-notebook-servers/jupyter-tensorflow-full/cuda.Dockerfile]: Fixed!

File: components/example-notebook-servers/jupyter-tensorflow/cpu.Dockerfile (hunk 11 above)
Comment (apo-ger) [reply components/example-notebook-servers/jupyter-tensorflow/cpu.Dockerfile]: Fixed!

File: components/example-notebook-servers/jupyter-tensorflow/cuda.Dockerfile (hunk 12 above)
Comment (apo-ger) [reply components/example-notebook-servers/jupyter-tensorflow/cuda.Dockerfile]: Fixed!

File: components/example-notebook-servers/rstudio-tidyverse/Dockerfile (hunk 14 above)
Comment (apo-ger) [reply components/example-notebook-servers/rstudio-tidyverse/Dockerfile]: Fixed!

File: components/example-notebook-servers/rstudio/Dockerfile (hunk 15 above)
Comment (apo-ger) [reply components/example-notebook-servers/rstudio/Dockerfile]: Fixed!

**Comment by kimwnasptd** [comment]
I'm getting the following error:
```
Building Notebook-controller image...

make docker-build -C notebook-controller
make[1]: Entering directory '/home/kimwnasptd/Code/git/kubeflow/components/notebook-controller'
Makefile:78: *** missing separator.  Stop.
make[1]: Leaving directory '/home/kimwnasptd/Code/git/kubeflow/components/notebook-controller'
make: *** [Makefile:18: build-all] Error 2
```

I think there's an extra `|` char at the beginning of that line.

**Comment by apo-ger** [comment]
> 



> I'm getting the following error:
> 
> ```
> Building Notebook-controller image...
> 
> make docker-build -C notebook-controller
> make[1]: Entering directory '/home/kimwnasptd/Code/git/kubeflow/components/notebook-controller'
> Makefile:78: *** missing separator.  Stop.
> make[1]: Leaving directory '/home/kimwnasptd/Code/git/kubeflow/components/notebook-controller'
> make: *** [Makefile:18: build-all] Error 2
> ```
> 
> I think there's an extra `|` char at the beginning of that line.

@kimwnasptd I removed the misplaced char.

**Comment by kimwnasptd** [comment]
I almost forgot. We'll need to update the workflow that then builds all components at once https://github.com/kubeflow/kubeflow/blob/master/.github/workflows/all_components_docker_publish.yaml.

@apo-ger can you also change that workflow to 
1. use the Makefile under `components/Makefile`
2. take the value from `releasing/version/VERSION` (we do that already)
3. use that as the tag when building all images

**Comment by apo-ger** [comment]
@kimwnasptd 
> Thanks! There's one last item, let's update our workflow so that it uses this makefile. https://github.com/kubeflow/kubeflow/blob/master/.github/workflows/nb_server_docker_publish.yaml#L17
> 
> Specifically, lets:
> 
> 1. Have only 4 steps in the that workflow
> 2. The existing `Checkout` step
> 3. The existing `Login to DockerHub` step
> 4. A step for building all images, just by using that Makefile, `docker-build-all`
> 5. A step for pushing the images only if `github.event_name == 'push'` https://github.com/kubeflow/kubeflow/blob/master/.github/workflows/nb_server_docker_publish.yaml#L39 that uses `docker-push-all`

Done!

> I almost forgot. We'll need to update the workflow that then builds all components at once https://github.com/kubeflow/kubeflow/blob/master/.github/workflows/all_components_docker_publish.yaml.
> 
> @apo-ger can you also change that workflow to
> 
> 1. use the Makefile under `components/Makefile`
> 2. take the value from `releasing/version/VERSION` (we do that already)
> 3. use that as the tag when building all images

Done!

**Comment by kimwnasptd** [comment]
All the tests pass, great work @apo-ger!

/lgtm
/approve

**Comment by thesuperzapper** [comment]
@apo-ger as I mentioned in https://github.com/kubeflow/kubeflow/issues/6560#issuecomment-1204566265, are you willing/interested to extend this awesome change to also push to `ghcr.io/kubeflow/kubeflow/notebook-servers/` (GHCR)?

I am happy to give you some pointers if you are, and need help!

**Pull request merged**